include/cache.h

```diff
@@ -13,59 +13,185 @@
 extern "C" {
 #endif
 
-void arch_dcache_flush(void *addr, size_t size);
-void arch_dcache_invd(void *addr, size_t size);
+/**
+ * Common operations for the caches
+ */
+#define K_CACHE_WB	BIT(0)
+#define K_CACHE_INVD	BIT(1)
+#define K_CACHE_WB_INVD	(K_CACHE_WB | K_CACHE_INVD)
+
+/**
+ *
+ * @brief Enable d-cache
+ *
+ * Enable the d-cache.
+ *
+ * @return N/A
+ */
+void arch_dcache_enable(void);
+
+/**
+ *
+ * @brief Disable d-cache
+ *
+ * Disable the d-cache.
+ *
+ * @return N/A
+ */
+void arch_dcache_disable(void);
+
+/**
+ *
+ * @brief Enable i-cache
+ *
+ * Enable the i-cache.
+ *
+ * @return N/A
+ */
+void arch_icache_enable(void);
+
+/**
+ *
+ * @brief Disable i-cache
+ *
+ * Disable the i-cache.
+ *
+ * @return N/A
+ */
+void arch_icache_disable(void);
+
+/**
+ *
+ * @brief Write-back / Invalidate / Write-back + Invalidate one level d-cache
+ *
+ * Write-back, Invalidate or Write-back + Invalidate one lever of the d-cache
+ *
+ * @param level	The cache level to act on
+ * @param op	Operation to perform (one of the K_CACHE_* operations)
+ *
+ * @retval 0		On success
+ * @retval -ENOTSUP	If the operation is not supported
+ */
+int arch_dcache_level(int level, int op);
+
+/**
+ *
+ * @brief Write-back / Invalidate / Write-back + Invalidate all d-cache
+ *
+ * Write-back, Invalidate or Write-back + Invalidate the whole d-cache.
+ *
+ * @param op	Operation to perform (one of the K_CACHE_* operations)
+ *
+ * @retval 0		On success
+ * @retval -ENOTSUP	If the operation is not supported
+ */
+int arch_dcache_all(int op);
 
 /**
  *
- * @brief Flush d-cache lines to main memory
+ * @brief Write-back / Invalidate / Write-back + Invalidate d-cache lines
  *
  * No alignment is required for either addr or size, but since
- * sys_cache_flush() iterates on the d-cache lines, a d-cache line alignment for
- * both is optimal.
+ * arch_dcache_range() iterates on the d-cache lines, a d-cache line alignment
+ * for both is optimal.
  *
  * The d-cache line size is specified either via the CONFIG_CACHE_LINE_SIZE
  * kconfig option or it is detected at runtime.
  *
- * @param addr the pointer to start the multi-line flush
- * @param size the number of bytes that are to be flushed
+ * @param addr	The pointer to start the multi-line operation
+ * @param size	The number of bytes that are to be acted on
+ * @param op	Operation to perform (one of the K_CACHE_* operations)
  *
- * @return N/A
+ * @retval 0		On success
+ * @retval -ENOTSUP	If the operation is not supported
  */
-__syscall void sys_cache_flush(void *addr, size_t size);
+int arch_dcache_range(void *addr, size_t size, int op);
```

Comment:
What would be the correct interpretation of this function on a CPU cache controller which can perform an operation (for example `INVD`) on the full cache but not on a per-line basis? (The LEON3 processors is one such example.) I see two main interpretations:
1. Do the operation on the full i-cache or d-cache. `return 0`.
or
2. Do no operation. `return -ENOTSUP`.

Reply:
I'd say that this operation would return `-ENOTSUP` but you can still use `arch_dcache_all / arch_icache_all` in that case (and that should return `0`).

Reply:
Thanks for the clarification. It makes sense for the low-level API.

One use case I had in mind was invalidating/flushing buffers in DMA based peripheral drivers on systems which do not have cache coherency between CPU and DMA. If we consider a scenario where the driver wants to invalidate cache state for 10 small buffers located at various places in RAM because a DMA read operation has finished. Then if the CPU has the capability to operate per-line, then it is probably best to do 10 individual `arch_dcache_range()` in a loop over DMA buffer desriptors. On a CPU which lacks the per-line invalidation capability, the right thing to do is likely a single invalidation with `arch_dcache_all()`.

That use case is all fine with the suggested API.